Design-Docs/tools/scs-compliance-check.py

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python3
+# vim: set ts=4 sw=4 et:
+#
+# scs-compliance-check.py
+#
+# (c) Eduard Itrich <eduard@itrich.net>
+# (c) Kurt Garloff <kurt@garloff.de>
+# SPDX-License-Identifier: Apache-2.0
+
+"""Master SCS compliance checker
+reads SCS certification requirements from e.g. scs-compatible.yaml
+and performs all the checks for the specified level and outputs a
+verdict from all tests (which is reflected in the exit code)
+The tests require the OpenStack SDK (and in the future probably
+also k8s python bindings) to be installed and access to IaaS
+(for the iaas layer tests) via configure clouds/secure.yaml
+which are passed in OS_CLOUD (or --os-cloud cmdline param).
+In the future als access to a cluster with KUBECONFIG pointing
+to a working file.
+The goal is to work without any special admin privileges.
+(If we find things that can't be tested as normal user, we
+would split these tests out.)
+"""
+
+import os
+import sys
+import getopt
+# import time
+import datetime
+import subprocess
+import copy
+import yaml
+
+
+def usage():
+    "Output usage information"
+    print("Usage: scs-compliance-check.py [options] compliance-spec.yaml layer [layer [layer]]")
+    print("Options: -v/--verbose: More verbose output")
+    print(" -q/-quiet: Don't output anythin but errors")
+    print(" -s/--single-layer: Don't perform require checks for dependantr layers")
+    print(" -d/--date YYYY-MM-DD: Check standards valid on specified date instead of today")
+    print(" -V/--version VERS: Force version VERS of the standard (instead of deriving from date)")
+    print(" -c/--os-cloud CLOUD: Use specified cloud env (instead of OS_CLOUD env var)")
+    print(" -o/--output path: Generate report of compliance check under given path")
+
+
+def is_valid_standard(now, stable, obsolete):
+    "Check if now is after stable and not after obsolete"
+    if not stable:
+        return False
+    if now < stable:
+        return False
+    if obsolete and now > obsolete:
+        return False
+    return True
+
+
+def run_check_tool(executable, args, verbose=False, quiet=False):
+    "Run executable and return exit code"
+    exe = [executable, ]
+    if args:
+        exe.extend(args.split(" "))
+    # print(f"{exe}")
+    # compl = subprocess.run(exe, capture_output=True, text=True, check=False)
+    compl = subprocess.run(exe, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
```

Comment:
Just an idea which might lead to a separate issue: Instead of using subprocess, import the files as libraries dynamically? E.g. like this http://web.archive.org/web/20120315061111/http://www.diveintopython.net/functional_programming/dynamic_import.html

Reply:
What if a test (in the future) won't be a Python script but rather some binary executable?

Reply:
I would keep the design generic and have the ability here to run any executable, e.g. also shell scripts.